File: camerax_android/lib/src/ml/mlkit_analyzer_impl.dart
import 'package:camerax_android/src/camerax_api.g.dart';
import 'package:camerax_android/src/common.dart';
import 'package:camerax_android/src/core.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

import 'barcode.dart';
import 'face.dart';
import 'interfaces.dart';

final class MlKitAnalyzerResultImpl extends MlKitAnalyzerResult {
  final MlKitAnalyzerResultApi api;

  MlKitAnalyzerResultImpl.internal(this.api, {required super.timestamp})
    : super.impl();

  @override
  Future<T?> getValue<T>(Detector<T> detector) {
    if (detector is BarcodeScannerImpl) {
      return api
          .getValue1(detector.barcodeScannerApi)
          .then((e) => e?.map((e1) => e1.impl).toList() as T?);
    } else if (detector is FaceDetectorImpl) {
      return api
          .getValue2(detector.faceDetectorApi)
          .then((e) => e?.map((e1) => e1.impl).toList() as T?);
    } else {
      throw TypeError();
    }
  }

  @override
  Future<Object?> getThrowable<T>(Detector<T> detector) {
    if (detector is BarcodeScannerImpl) {
      return api.getThrowable1(detector.barcodeScannerApi).then((e) => e?.impl);
    } else if (detector is FaceDetectionImpl) {
      return api.getThrowable2(detector.faceDetectorApi).then((e) => e?.impl);
    } else {
      throw TypeError();
    }
  }
}

final class MlKitAnalyzerImpl extends MlKitAnalyzer
    with ImageAnalysisAnalyzerImpl {
  @override
  final MlKitAnalyzerApi api;

  MlKitAnalyzerImpl.internal(this.api) : super.impl();

  factory MlKitAnalyzerImpl({
    required List<Detector> detectors,
    required ImageAnalysisCoordinateSystem targetCoordinateSystem,
    required Consumer<MlKitAnalyzerResult> consumer,
  }) {
    final api = MlKitAnalyzerApi(
      detectors1: detectors
          .whereType<BarcodeScanner>()
          .map((e) => e.api)
          .toList(),
      detectors2: detectors
          .whereType<FaceDetector>()
          .map((e) => e.api)
          .toList(),
      targetCoordinateSystem: targetCoordinateSystem.api,
      consumer: consumer.mlKitAnalyzerResultConsumerApi,
    );
    return MlKitAnalyzerImpl.internal(api);
  }
}

extension MlKitAnalyzerResultApiX on MlKitAnalyzerResultApi {
  MlKitAnalyzerResult get impl {
    return MlKitAnalyzerResultImpl.internal(this, timestamp: timestamp);
  }
}

File: camerax_android/android/src/main/kotlin/dev/zeekr/camerax_android/ml/MlKitAnalyzerImpl.kt
package dev.zeekr.camerax_android.ml

import androidx.camera.mlkit.vision.MlKitAnalyzer
import androidx.core.content.ContextCompat
import com.google.mlkit.vision.barcode.BarcodeScanner
import com.google.mlkit.vision.barcode.common.Barcode
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceDetector
import dev.zeekr.camerax_android.*
import dev.zeekr.camerax_android.common.MlKitAnalyzerResultConsumer
import dev.zeekr.camerax_android.common.api
import dev.zeekr.camerax_android.core.impl

class MlKitAnalyzerImpl(private val registrar: CameraXApiPigeonProxyApiRegistrar) :
    PigeonApiMlKitAnalyzerProxyApi(registrar) {
    override fun pigeon_defaultConstructor(
        detectors1: List<BarcodeScanner>,
        detectors2: List<FaceDetector>,
        targetCoordinateSystem: ImageAnalysisCoordinateSystemApi,
        consumer: MlKitAnalyzerResultConsumer
    ): MlKitAnalyzer {
        val executor = ContextCompat.getMainExecutor(registrar.context)
        return MlKitAnalyzer(detectors1 + detectors2, targetCoordinateSystem.impl, executor, consumer)
    }

    class ResultImpl(registrar: CameraXApiPigeonProxyApiRegistrar) : PigeonApiMlKitAnalyzerResultProxyApi(registrar) {
        override fun timestamp(pigeon_instance: MlKitAnalyzer.Result): Long {
            return pigeon_instance.timestamp
        }

        override fun getValue1(pigeon_instance: MlKitAnalyzer.Result, detector: BarcodeScanner): List<Barcode>? {
            return pigeon_instance.getValue(detector)
        }

        override fun getValue2(pigeon_instance: MlKitAnalyzer.Result, detector: FaceDetector): List<Face>? {
            return pigeon_instance.getValue(detector)
        }

        override fun getThrowable1(pigeon_instance: MlKitAnalyzer.Result, detector: BarcodeScanner): List<Any?>? {
            return pigeon_instance.getThrowable(detector)?.api
        }

        override fun getThrowable2(pigeon_instance: MlKitAnalyzer.Result, detector: FaceDetector): List<Any?>? {
            return pigeon_instance.getThrowable(detector)?.api
        }
    }
}